server/remote_asset/fetch_server/fetch_server.go

```diff
@@ -111,31 +121,43 @@ func (p *FetchServer) FetchBlob(ctx context.Context, req *rapb.FetchBlobRequest)
 			}
 			blobDigest := &repb.Digest{
 				Hash: fmt.Sprintf("%x", sha256),
-				// We don't know the digest size (yet), but we have to specify
-				// something here or the bytestream API will return an
-				// InvalidArgument error. The bytestream server doesn't actually
-				// check this value; it just needs something greater than 0.
-				//
-				// Note, the value we specify here is used as the max size of
-				// each chunk sent back in the stream, so want something
-				// relatively large here or else the streaming will be too slow.
-				SizeBytes: bytestreamReadBufSize,
+				// The digest size is unknown since the client only sends up
+				// the hash. We can look up the size using the Metadata API,
+				// which looks up only using the hash, so the size we pass here
+				// doesn't matter.
+				SizeBytes: -1,
 			}
 			expectedSHA256 = blobDigest.Hash
 			cacheRN := digest.NewCASResourceName(blobDigest, req.GetInstanceName())
 
-			// TODO: Find a way to get the correct digest size without
-			// downloading the full blob from cache.
 			log.CtxInfof(ctx, "Looking up %s in cache", blobDigest.Hash)
-			c := &ioutil.Counter{}
 
-			if err := cachetools.GetBlob(ctx, p.env.GetByteStreamClient(), cacheRN, c); err != nil {
-				log.CtxInfof(ctx, "FetchServer failed to get %s from cache: %s", expectedSHA256, err)
+			// Lookup metadata to get the correct digest size to be returned to
+			// the client.
+			md, err := p.env.GetCache().Metadata(ctx, cacheRN.ToProto())
+			if err != nil {
+				log.CtxInfof(ctx, "FetchServer failed to get metadata for %s: %s", expectedSHA256, err)
 				continue
 			}
-
-			blobDigest.SizeBytes = c.Count() // set the actual correct size.
-			log.CtxInfof(ctx, "FetchServer successfully read %s from cache", digest.String(blobDigest))
+			blobDigest.SizeBytes = md.SizeBytes
+			// Even though we successfully fetched metadata, we need to renew
+			// the cache entry using FindMissingBlobs to ensure that it doesn't
+			// expire by the time the client requests it from cache.
+			cas := p.env.GetContentAddressableStorageClient()
+			req := &repb.FindMissingBlobsRequest{
+				InstanceName: req.GetInstanceName(),
+				BlobDigests:  []*repb.Digest{blobDigest},
+			}
+			res, err := cas.FindMissingBlobs(ctx, req)
```

Comment:
Basically the same but cache.Contains can also update LRU entries - unclear if either is clearer 🤷‍♀️